File: Python/leetcode_46.py
from typing import List


def permute(nums: List[int]) -> List[List[int]]:
    result: List[List[int]] = []

    def backtrack(path: List, used: set):
        if len(path) == len(nums):
            result.append(path[:])
            return
        for i in nums:
            if i in used:
                continue
            path.append(i)
            used.add(i)
            backtrack(path, used)
            path.pop()
            used.remove(i)
    backtrack([], set())

    return result

File: JavaScript/leetcode_46.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function (nums) {
    let result = [];
    /**
     * @param {number[]} path
     * @param {Set} used
     */
    function backtrack(path, used) {
        if (path.length === nums.length) {
            result.push([...path]);
            return;
        }
        for (let index = 0; index < nums.length; index++) {
            if (used.has(nums[index])) continue;
            path.push(nums[index]);
            used.add(nums[index]);
            backtrack(path, used);
            path.pop();
            used.delete(nums[index]);
        }
    }
    backtrack([], new Set());
    return result;
};
